clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchManagerMetrics.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.metrics.Gauge;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.Avg;
+import org.apache.kafka.common.metrics.stats.Max;
+import org.apache.kafka.common.metrics.stats.Meter;
+import org.apache.kafka.common.metrics.stats.Min;
+import org.apache.kafka.common.metrics.stats.Value;
+import org.apache.kafka.common.metrics.stats.WindowedCount;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The {@link FetchManagerMetrics} class provides wrapper methods to record lag, lead, latency, and fetch metrics.
+ * It keeps an internal ID of the assigned set of partitions which is updated to ensure the set of metrics it
+ * records matches up with the topic-partitions in use.
+ */
+class FetchManagerMetrics {
+
+    private final Metrics metrics;
+    private final FetcherMetricsRegistry metricsRegistry;
+    private final Sensor bytesFetched;
+    private final Sensor recordsFetched;
+    private final Sensor fetchLatency;
+    private final Sensor recordsFetchLag;
+    private final Sensor recordsFetchLead;
+
+    private int assignmentId = 0;
+    private Set<TopicPartition> assignedPartitions = Collections.emptySet();
```

Comment:
This is not introduced by this PR, but I wonder if it is better i.e. stricter, to just go over all the metrics matching a topic-partition name matcher against the `newAssignedPartitions` and remove if necessary, than remembering the old assignment and just loop over these, given that `maybeUpdateAssignment` now can be triggered from different callers.

Reply:
Ditto with remembering the `assignmentId`.

Reply:
I'm not totally understanding the concept, sorry. Are you suggesting we do something like this:

```java
    void maybeUpdateAssignment(SubscriptionState subscription) {
        Set<TopicPartition> partitions = subscription.assignedPartitions();

        for (MetricName metricName : metrics.metrics().keySet()) {
            TopicPartition tp = parseTopicPartitionFromMetricName(metricName);
            
            if (!partitions.contains(tp)) {
                metrics.removeSensor(partitionLagMetricName(tp));
                metrics.removeSensor(partitionLeadMetricName(tp));
                metrics.removeMetric(partitionPreferredReadReplicaMetricName(tp));
            }
        }

        for (TopicPartition tp : partitions) {
            MetricName metricName = partitionPreferredReadReplicaMetricName(tp);
            metrics.addMetricIfAbsent(
                    metricName,
                    null,
                    (Gauge<Integer>) (config, now) -> subscription.preferredReadReplica(tp, 0L).orElse(-1)
            );
        }
    }
```

Reply:
Yeah something like that (though we need to check if all metric names did include the partition in the same format), instead of trying to remember all the prev assigned partitions instead. The idea is that since now the assignment could be updated by different threads, we may lose track of the full history of `assigned partitions` and hence would cause some dangling metrics in the registry never to be removed, or some metrics unnecessarily removed, by just relying on the new assignment and always loop over all metrics to trim them seems safer.

Reply:
Anyways, we can defer this change for future changes since it's not introduced by this PR.